src/krb5/_set_password.pyi

```diff
@@ -1,31 +1,112 @@
+    - `history`    - Number of passwords that this system remembers
+    - `max_age`    - Maximum password age in 100 nanosecond units
+    - `min_age`    - Minimum password age in 100 nanosecond units
+
+    The only known property flag is `COMPLEX` which means that the password must
+    meet certain character variety and not contain the user's name.
+    To convert `max_age` and `min_age` to seconds, divide them by 10,000,000.
+    """
+
+    properties: ADPolicyInfoProp
+    min_length: int
+    history: int
+    max_age: int
+    min_age: int
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> "ADPolicyInfo":
+        """Decode AD policy result from byte string
+
+        Args:
+            data: Serialized AD policy `server_response`
+
+        Returns:
+            ADPolicyInfo: Decoded AD policy result strcture
+
+        Raises:
+            ValueError: Invalid data length or signature not 0x0000
+        """
+    @classmethod
+    def to_bytes(cls, policy: "ADPolicyInfo") -> bytes:
+        """Reverses the `from_bytes` operation
+
+        Args:
+            policy: AD policy result structure
+
+        Returns:
+            bytes: Serialized AD policy result byte string
+        """
+
+class SetPasswordResultCode(enum.IntEnum):
+    SUCCESS = 0
+    MALFORMED = 1
+    HARDERROR = 2
+    AUTHERROR = 3
+    SOFTERROR = 4
+
 class SetPasswordResult(typing.NamedTuple):
     """The result returned by :meth:`set_password()` and
     :meth:`set_password_using_ccache()`.
 
-    The `result_code` and `result_code_string` is the library response:\n
-    KRB5_KPASSWD_SUCCESS   (0) - Success\n
-    KRB5_KPASSWD_MALFORMED (1) - Malformed request error\n
-    KRB5_KPASSWD_HARDERROR (2) - Server error\n
-    KRB5_KPASSWD_AUTHERROR (3) - Authentication error\n
-    KRB5_KPASSWD_SOFTERROR (4) - Password change rejected\n
+    The `result_code` and `result_code_string` are the pure library responses:
+    - SUCCESS   (0) - Success
+    - MALFORMED (1) - Malformed request error
+    - HARDERROR (2) - Server error
+    - AUTHERROR (3) - Authentication error
+    - SOFTERROR (4) - Password change rejected
 
-    The `result_string` is a server protocol response that may contain useful
+    The `server_response` is a server protocol message that may contain useful
     information about password policy violations or other errors.
+    Depending on `kpasswd` implementation, it may be returned as:\n
+    - decoded UTF-8 string (MIT KDC)
+    - decoded `ADPolicyInfo` (Active Directory Policy Information)
+    - raw bytes (if unable to decode)
+
+    When (and only when) the server response is exactly 30 bytes long starting
+    with `0x0000`, it is assumed to be `ADPolicyInfo`.
+    All other cases are first decoded as UTF-8 string and even if this fails,
+    the raw bytes are returned as `server_response`.
+
+    See `ADPolicyInfo` for more information.
     """
 
-    result_code: int
+    result_code: SetPasswordResultCode
```

Comment:
We need to be careful about returning an enum, if a value is set here which we haven't added to the enum it will fail to cast

```python
import enum

class MyEnum(enum.IntEnum):
    FOO = 1

MyEnum(2)
# ValueError: 2 is not a valid MyEnum
```

See [NameType](https://github.com/jborean93/pykrb5/blob/e63460989f1085ba41739c5651f29601ed3209dd/src/krb5/_principal.pyx#L253-L262) for how to deal with missing members so this will still work.

Reply:
check my "best effort" attempt